matlab/src/Util.cpp

```diff
@@ -330,156 +331,106 @@ IceMatlab::getProtocolVersion(mxArray* p, Ice::ProtocolVersion& v)
     getMajorMinor(p, v.major, v.minor);
 }
 
+namespace
+{
+    template<size_t N> mxArray* createMatlabException(const char* typeId, std::array<mxArray*, N> params)
+    {
+        string className = replace(string{typeId}.substr(2), "::", ".");
+        mxArray* ex;
+        mexCallMATLAB(1, &ex, static_cast<int>(N), params.data(), className.c_str()); // error is fatal
+        return ex;
+    }
+
+    template<size_t N> mxArray* createMatlabExceptionWithTrap(const char* typeId, std::array<mxArray*, N> params)
+    {
+        string className = replace(string{typeId}.substr(2), "::", ".");
+        mxArray* ex;
+        mexCallMATLABWithTrap(1, &ex, static_cast<int>(N), params.data(), className.c_str()); // keep going on error
+        return ex;
+    }
+}
+
 mxArray*
 IceMatlab::convertException(const std::exception_ptr exc)
 {
-    mxArray* ex;
+    const char* const localExceptionTypeId = "::Ice::LocalException";
+    mxArray* result;
 
     try
     {
         rethrow_exception(exc);
     }
-    catch (const Ice::LocalException& iceEx)
+    // We need to catch and convert:
+    // - local exceptions thrown from MATLAB code for which we provide a convience constructor (e.g. MarshalException)
+    // - local exceptions that define extra properties we want to expose to MATLAB users (e.g. ObjectNotExistException
+    // via its base class, RequestFailedException)
+    catch (const Ice::RequestFailedException& e)
     {
-        string typeId{iceEx.ice_id()};
-        //
-        // The exception ID uses single colon separators.
-        //
-        auto id = typeId.substr(2); // Remove leading "::" from type ID
-        id = replace(id, "::", ":");
-
-        auto cls = typeId.substr(2); // Remove leading "::" from type ID
-        cls = replace(cls, "::", ".");
-
-        mxArray* params[10];
-        params[0] = createStringFromUTF8(id);
-        int idx = 2;
-        string msg = typeId; // Use the type ID as the default exception message
+        // The *NotExist exceptions are thrown only from the C++ code. They don't have a convenience constructor, but
+        // they have extra properties.
+        string errID = replace(string{e.ice_id()}.substr(2), "::", ":");
```

Comment:
```suggestion
        string errID = replace(string{e.ice_id()}.substr(2), "::", ".");
```

Is the replacement value `:` correct?

Reply:
Yes, it's correct. errID use the `:` separator, not `.`.

Reply:
The language separator for classes is '.' and Slice is '::', why use a different separator for Error ids, is this common to all the mappings now?

Reply:
This separator comes from MATLAB:
https://www.mathworks.com/help/matlab/ref/mexception.html#bq9tdlq-1
(under the `errID` section... couldn't link directly to it).

Reply:
> The error identifier includes one or more component fields and a mnemonic field. Fields must be separated with colon. For example, an error identifier with a component field component and a mnemonic field mnemonic is specified as 'component:mnemonic'.